### PR TITLE
Remove watching Endpoints of Headless Services

### PR DESCRIPTION
**Pull request by rcarrillocruz**

### CODE REVIEW

File: pkg/network/proxy/proxy.go [hunk 1]
```diff
@@ -312,49 +314,58 @@ func (proxy *OsdnProxy) OnEndpointsAdd(ep *corev1.Endpoints) {
 	proxy.Lock()
 	defer proxy.Unlock()
 
-	pep := &proxyEndpoints{ep, proxy.endpointsBlocked(ep)}
-	proxy.allEndpoints[ep.UID] = pep
-	if !pep.blocked {
-		proxy.baseProxy.OnEndpointsAdd(ep)
+	labels := ep.GetLabels()
+	if _, ok := labels[IsHeadLessService]; ok {
+		pep := &proxyEndpoints{ep, proxy.endpointsBlocked(ep)}
```
Comment (danwinship) [inline pkg/network/proxy/proxy.go]: Don't indent the whole function body; return early in the no-op case instead.

And the check is backwards anyway isn't it? (We want to process the endpoints if it *doesn't* have the annotation, and skip processing if it does.)

File: pkg/network/proxy/proxy.go [hunk 2]
```diff
@@ -312,49 +314,58 @@ func (proxy *OsdnProxy) OnEndpointsAdd(ep *corev1.Endpoints) {
...
-
-	switch {
-	case wasBlocked && !pep.blocked:
-		proxy.baseProxy.OnEndpointsAdd(ep)
-	case !wasBlocked && !pep.blocked:
-		proxy.baseProxy.OnEndpointsUpdate(old, ep)
-	case !wasBlocked && pep.blocked:
-		proxy.baseProxy.OnEndpointsDelete(ep)
+	labels := ep.GetLabels()
+	if _, ok := labels[IsHeadLessService]; ok {
```
Comment (danwinship) [inline pkg/network/proxy/proxy.go]: maybe have a helper function

File: pkg/network/proxy/proxy.go [hunk 3]
```diff
@@ -26,6 +26,8 @@ import (
 	"github.com/openshift/sdn/pkg/network/common"
 )
 
+const IsHeadLessService string = "service.kubernetes.io/headless"
```
Comment (danwinship) [inline pkg/network/proxy/proxy.go]: This is already exported upstream: `corev1.IsHeadlessService`

**Comment by squeed** [comment]
The upstream implementation of this applies the label selector directly to the watch, which means the services aren't even sent down the wire. This is definitely a big improvement, and will help ups because of the large services created for monitoring (e.g. node-exporter, openshift-sdn, etc).

The problem is that our watches are shared between kube-proxy and the network policy code. Do you know if network policy also uses endpoints? If nothing besides kube-proxy watches endpoints, then it would be safe for us to implement this in the same manner as upstream.

**Comment by squeed** [comment]
I did a quick scan of the codebase, and we don't use the Endpoint informer anywhere else. So I'd say it's safe to add the label selector directly.

**Comment by danwinship** [comment]
Hm... I hadn't considered that doing the filter at the informer level would also improve performance.

So maybe it would be better to do it that way. OK, sorry, @rcarrillocruz, I screwed up. You did it right the first time...

**Pull request merged**